Update tclap to version 1.4 for better compiler support

hjmjohnson opened the pull request:
Updates the vendored TCLAP from 1.2 to **1.4.0-rc2** (headers byte-identical to the upstream tag) and defaults SlicerExecutionModel to **C++17** (C++20/23 also accepted). The motivating win: TCLAP is `#include`d into *every* generated module, so its header warnings are our warnings — this drop deletes the deprecated `<strstream>` / header-level `using namespace std` code entirely, which is what unblocks `-Werror` across SEM consumers (Slicer, BRAINSTools, ANTs).

<details>
<summary>Why this matters more than a normal dependency bump</summary>

`GenerateCLP` emits `#include "tclap/CmdLine.h"` into every generated module
(`GenerateCLP/GenerateCLP.cxx:463`). TCLAP is header-only — it is **recompiled
as part of the consumer's code** in every CLI executable, not linked. A warning
in a TCLAP header is therefore one warning × every module × every translation
unit. That multiplier is why updating TCLAP measurably reduces warnings in
downstream code.

</details>

<details>
<summary>What changed (the warning sources removed)</summary>

Six legacy headers deleted: `ArgOld.h`, `MultiArgOld.h`, `ValueArgOld.h`,
`XMLOutput.h`, `XorHandler.h`, `ZshCompletionOutput.h`.

- `MultiArgOld.h` / `ValueArgOld.h` did `#include <strstream>` unconditionally
  and used `std::istrstream`. `<strstream>` has been deprecated since C++98 —
  modern compilers fire `-Wdeprecated-declarations`, and C++26 removes it.
- `StdOutput.h` had a header-level `using namespace std;` (namespace-leak /
  ADL antipattern, flagged by `clang-tidy google-build-using-namespace`).

Replacement: `sstream.h` now defaults to `<sstream>`
(`std::istringstream`/`std::ostringstream`); the `<strstream>` branch is only
reachable if `TCLAP_HAVE_STRSTREAM` is explicitly forced. This PR's CI confirms
the modern path is taken (`Performing Test TCLAP_HAVE_STRSTREAM - Failed` →
`TCLAP_HAVE_SSTREAM - Success`). The deprecated XOR handler is replaced by the
RAII-based `ArgGroup.h` / `ArgContainer.h` / `DeferDelete.h`.

</details>

<details>
<summary>The case for accepting</summary>

1. **Eliminates warnings at the source, in consumer code** — not by suppression.
   Because TCLAP is injected via `GenerateCLP`, the deprecation warnings lived
   inside compilation of *your* modules and could not be cleanly silenced.
   Deleting the offending headers stops the warnings being generated at all.
2. **Prerequisite for `-Werror`.** Unavoidable third-party-header warnings are
   what force a project to keep warnings non-fatal, letting real regressions
   hide in the noise. Clearing TCLAP's warnings unblocks fatal-warning builds.
3. **Future-proofs against hard-stops.** `<strstream>` is removed in C++26 and
   already gated behind deprecation in newer libstdc++/libc++; a toolchain bump
   would eventually turn old TCLAP's warnings into build *failures*.
4. **Low risk, honest diff.** The 26 headers are byte-identical to upstream
   `1.4.0-rc2`; only minimal build-glue differs (recorded by an in-tree sync
   marker). Upstream's ChangeLog states 1.2 programs "continue to compile out
   of the box," and the SEM test suite (15/15) compiles and runs generated
   parsers against the 1.4 API.
5. **Pays down maintenance debt.** Six dead headers and the autotools machinery
   are gone; future re-syncs are a file-level diff against a tagged release.

</details>

<details>
<summary>Verification</summary>

- Headers verified byte-identical to upstream tag `1.4.0-rc2` (commit `81b3d2a0`).
- Clean local build with **0 errors / 0 warnings** under both C++17 (default)
  and explicit `-DCMAKE_CXX_STANDARD=20`; **15/15** tests pass in both.
- CircleCI `itk-master_use_system_libraries-off` green on current HEAD.

</details>

<!--
provenance: claude-code session 2026-06-15
key_facts: vendored TCLAP == upstream tag 1.4.0-rc2 (commit 81b3d2a0); headers byte-identical; C++ default 17, valid 17/20/23; CI fix = git-track tclap/config.h.in past the tclap/.gitignore *.in glob
related_files: GenerateCLP/GenerateCLP.cxx:463 (emits #include tclap/CmdLine.h); tclap/include/tclap/sstream.h (sstream-default shim); tclap/CMakeLists.txt (sync marker); CMakeLists.txt:11,19 (C++ standard policy)
resync_recipe: diff tclap/ against upstream git tag 1.4.0-rc2; *.cmake.in, GenerateTCLAPConfig.cmake, config.h.in and SLICER_DISABLE_FROM_UPSTREAM blocks are Slicer-only glue
-->

hjmjohnson committed:
ENH: Update TCLAP to match master upstream branch

git clone https://git.code.sf.net/p/tclap/code

Manually merge changes from d6419ab71623821c5dd02be3bc2835577fa7d525

hjmjohnson commented:
related to #115

hjmjohnson commented:
@jcfr The patches for updating the docker image have already been posted #127 , but a new docker image needs to be generated and uploaded to dockerhub with the new version of cmake.  Then the WIP patch using the un-official docker image for CI can be removed from this PR.

hjmjohnson commented:
@jcfr closing stale pull requests.

jcfr commented:
Work to finalize this has been initiated. Now that the clang-tidy warnings have been integrated (https://github.com/Slicer/SlicerExecutionModel/pull/135) this will come next.

hjmjohnson commented:
@jcfr Thanks.  I'm just doing house cleaning.  I have too many open issues and pull requests that are slowing down more important work.  After 3 years, this is just not all that important to me. :).

Thank you for your continued support and efforts.

hjmjohnson committed:
ENH: Updates to tclap branch 1.4

This is from
  https://git.code.sf.net/p/tclap/code
  commit 81b3d2a0c47895c22e9bb8c577f5ab521f76e5d2
  branch 1.4 of tclap

Make virtual functions calls from constructor explicit.

Calling virtual functions from the ctor is safe in C++, but can be
confusing. This fixes clang-tidy warnings.

Fixes clan-tidy warnings (https://sourceforge.net/p/tclap/patches/18/).

Allow const Constraints in constructors

Constraints are only used in const-safe ways, so allow a caller
to pass a const Constraint.

It also allows static construction of constant constraints.

Add missing call to check visitor in Unlabeled args.

Fixed an error in packaging preventing proper discovery using CMake `find_package`

Added packaging using CPack and discovery using pkg-config.

Allowed to disable building of examples and unittests. And docs (BUILD_DOC used to have no effect, it is not the case since now.)

Remove confusing statement about not supplying flag argument.

This is more hurtful than helpful, flag doesn't have to be specified
(our docs even have this example), but flag or name does.

Fix compilation issue with C++20 and gcc-11.1

Fix some (harmless) warnings from Visual Studio.

VS issues warnings for `conversion from 'size_t' to 'int', possible
loss of data`. Of course since these are all short strings, it's not
an issue - but better to get rid of the warnings so we don't miss
anything else.

Remove instructions to use cmake --install . --prefix=..., it appears flaky.

It doesn't appear to work with my custom doc install targets, they
still revert to using the default value of CMAKE_INSTALL_PREFIX
(probably because the variable is expanded during configuration, not
install phase - but that's a guess).

Added docs to install targets.

There appears to be no standard way to install docs with CMake, so
I've setup paths for win and *NIX to the best of my ability.

Include generated docs in release tarball.

Also, make sure that release script doesn't have to be run from the
top level tclap directory.

Explicitly initiailze member variables in init-list.

clang-format on unittest files. No functional change.

Fixed typos.

Fix compilation issue on Windows.

Use ptrdiff_t (C++ type) instead of ssize_t (UNIX type).

Fix a bug where help output would crash if the program name was 3 chars.

https://sourceforge.net/p/tclap/patches/17/

Also added some very basic support for unit tests. We can look at pulling in some more robust unit testing framework later on, but for now keeping it simple to avoid bloat.

In addition, consider only platform specific path separators.

Script to copy files for release.

Make a note that CMake < 3.15 needs plaform specific installer.

Clarify that --prefix flag goes on the --install command.

--In preparation for release of v1.4.
  Upload docs for 1.4 to both 1.4 and root.
  Run clang-format on all files in prep. for 1.4 release.
  Clean the news file for 1.4 release.
  Note current maintainer in authors file.
  Update ChangeLog for 1.4 release.
  Cleaup manual.xml for 1.4 release candidate.

Update Google copyright stansa.

Fix compilation error on MacOS

operator>> has to be declared before inclusion of Arg.h (CmdLine.h).

Update copyright notices and version info.

Correct/clarify years and reorder in reverse chronological order in
the manual. Added note that this manual is for v1.4 since we'll soon
have two versions "live".

HTML manual is now built by cmake, no need to maintain it manually.

Make a note of it, and remove the checked in generated file.

Add example for how to use a pair as a custom type.

https://sourceforge.net/p/tclap/support-requests/2/

Update generated documentation for 1.4.

Echo commands from upload script.

To avoid myself getting confused when the right $USER is not set.

Add documentation for ExitException.

Looks like it may be useful for some users to catch this instead of
having TCLAP terminate the program. Also clarify that ArgException
doesn't catch *any* exception.

https://sourceforge.net/p/tclap/bugs/29/

Add missing expected test output from [96fd5ea].

Add some generated file types to gitignore (and sort it).

Check arguments vector is not empty before accessing first element.

Some users use TCLAP to parse messages sent between long-running
programs by puting arguments in a std::vector<std::string> and use
CmdLine::parse to parse it. Passing an empty vector to parse results
in a segmentation violation (due to the fact that we always expect to
be able to pop the first element). Change the behaviour to throw a
more user friendly error message.

https://sourceforge.net/p/tclap/bugs/30/

Revert dd2c7b6faa13ad74bce2bfe26e5483e62186ed6c that was pushed upstream by mistake.

Don't catch exceptions by value.

Exceptions are polymorphic, catching by value may loose information.

fix sign-conversion warnings

ENH: Adding travis build enviornment for CI testing

CI integration for many common platforms.
- Unix based platforms included
- Many common unix compilers included

Allow travis building of all branches

ENH: Adding my.cdash.org reporting site for testing.

To generate a report of the build & testing environment.

```
mkdir -p tclap-code-build
cd tclap-code-build
cmake ../tclap-code
make Experimental # or ninja Experimental or ...
```

Then open https://my.cdash.org/index.php?project=tclap to see the resulting report.

ENH: Explicitly add LANGUAGES CXX for project.

STYLE: Convert CMake-language commands to lower case

Ancient CMake versions required upper-case commands.  Later command names
became case-insensitive.  Now the preferred style is lower-case.

STYLE: Remove CMake-language block-end command arguments

Ancient versions of CMake required else(), endif(), and similar block
termination commands to have arguments matching the command starting the block.
This is no longer the preferred style.

ENH: enable_testing at top level

You need to move the enable_testing() call to be before you do add_subdirectory(tests).

https://stackoverflow.com/questions/30250494/ctest-not-detecting-tests

Clean up handling of locally allocated objects.

1) Just keep a default StdOutput around so we don't have to track if
it was set by user.

2) Generalize handling of deleting allocated objects when CmdLine is
destroyed.

Updated TODOs

Cleanup some simple setters.

Inline into the class if they are only one line and make them const
whenever possible.

hjmjohnson committed:
ENH: Default to C++17 and allow C++20/23 for SlicerExecutionModel

TCLAP 1.4 headers compile cleanly under modern standards and the
upstream 1.4.0-rc2 drop explicitly fixed C++20/gcc-11 builds. Keep
the SlicerExecutionModel default at C++17 while widening the valid
standards list from 11/14/17 to 17/20/23 so downstreams may opt up.
